x/stake/cross_stake/cross_stake.go

```diff
@@ -0,0 +1,356 @@
+	default:
+		panic("Unknown cross stake ack package type")
+	}
+	if err != nil {
+		panic(err)
+	}
+
+	return result
+}
+
+func (app *CrossStakeApp) ExecuteFailAckPackage(ctx sdk.Context, payload []byte) sdk.ExecuteResult {
+	if len(payload) == 0 {
+		app.stakeKeeper.Logger(ctx).Info("receive cross stake fail ack package")
+		return sdk.ExecuteResult{}
+	}
+
+	pack, err := DeserializeCrossStakeFailAckPackage(payload)
+	if err != nil {
+		app.stakeKeeper.Logger(ctx).Error("unmarshal cross stake fail ack package error", "err", err.Error(), "package", string(payload))
+		panic("unmarshal cross stake fail ack package error")
+	}
+
+	var result sdk.ExecuteResult
+	switch p := pack.(type) {
+	case types.CrossStakeDistributeRewardSynPackage:
+		result, err = app.handleDistributeRewardRefund(ctx, p)
+	case types.CrossStakeDistributeUndelegatedSynPackage:
+		result, err = app.handleDistributeUndelegatedRefund(ctx, p)
+	default:
+		panic("Unknown cross stake fail ack package type")
+	}
+	if err != nil {
+		panic(err)
+	}
+
+	return result
+}
+
+func (app *CrossStakeApp) ExecuteSynPackage(ctx sdk.Context, payload []byte, relayFee int64) sdk.ExecuteResult {
+	app.stakeKeeper.Logger(ctx).Info("receive cross stake syn package")
+	pack, err := DeserializeCrossStakeSynPackage(payload)
+	if err != nil {
+		app.stakeKeeper.Logger(ctx).Error("unmarshal cross stake sync claim error", "err", err.Error(), "claim", string(payload))
+		panic("unmarshal cross stake claim error")
+	}
+
+	var result sdk.ExecuteResult
+	switch p := pack.(type) {
+	case types.CrossStakeDelegateSynPackage:
+		result, err = app.handleDelegate(ctx, p, relayFee)
+	case types.CrossStakeUndelegateSynPackage:
+		result, err = app.handleUndelegate(ctx, p, relayFee)
+	case types.CrossStakeRedelegateSynPackage:
+		result, err = app.handleRedelegate(ctx, p, relayFee)
+	default:
+		panic("Unknown cross stake syn package type")
+	}
+	if err != nil {
+		panic(err)
+	}
+
+	return result
+}
+
+func (app *CrossStakeApp) handleDelegate(ctx sdk.Context, pack types.CrossStakeDelegateSynPackage, relayFee int64) (sdk.ExecuteResult, error) {
+	sideChainId := app.stakeKeeper.DestChainName
+	if scCtx, err := app.stakeKeeper.ScKeeper.PrepareCtxForSideChain(ctx, sideChainId); err != nil {
+		return sdk.ExecuteResult{}, err
+	} else {
+		ctx = scCtx
+	}
+
+	delAddr := types.GetStakeCAoB(pack.DelAddr[:], types.DelegateCAoBSalt)
+	validator, found := app.stakeKeeper.GetValidator(ctx, pack.Validator)
+	if !found || validator.Jailed {
+		var sdkErr sdk.Error
+		var errCode uint8
+		if !found {
+			sdkErr = types.ErrNoValidatorFound(types.DefaultCodespace)
+			errCode = CrossStakeErrValidatorNotFound
+		} else {
+			sdkErr = types.ErrValidatorJailed(types.DefaultCodespace)
+			errCode = CrossStakeErrValidatorJailed
+		}
+		ackPack := types.NewCrossStakeDelegationAckPackage(&pack, types.CrossStakeTypeDelegate, errCode)
+		ackPack.Amount = bsc.ConvertBCAmountToBSCAmount(ackPack.Amount.Int64())
```

Comment:
Why not do `ConvertBCAmountToBSCAmount` inside `NewCrossStakeDelegationAckPackage`?

Reply:
Done